onnxruntime/contrib_ops/cuda/moe/moe.cc

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "moe.h"
+
+using namespace onnxruntime::cuda;
+using namespace ::onnxruntime::common;
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+#define REGISTER_KERNEL_TYPED(T)                                                                                     \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(MoE, kMSDomain, 1, T, kCudaExecutionProvider,                                        \
+                                (*KernelDefBuilder::Create()).TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+                                MoE<T>);
```

Comment:
Could you test whether it is able to save output to the buffer of input? 
If so, we can add MayInPlace to reduce some memory usage.

Reply:
added